### PR TITLE
CSI 3418 | add the option to developers to do generation in thier local env

### PR DESCRIPTION
**Pull request by matancarmeli7**
Signed-off-by: matancarmeli7 <matan.carmeli7@gmail.com>
this is also fixes ticket CSI-3252

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -61,19 +61,43 @@ rm -rf $$TMP_DIR ;\
 endef
 
 # custom
+run_unit_tests_image=docker run --rm -e KUSTOMIZE=$(KUSTOMIZE) -v "${PWD}":/go/src/github.com/IBM/ibm-block-csi-operator -t operator-unittests
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply Makefile:64]: sure

File: build/ci/Dockerfile.unittest [hunk 2]
```diff
@@ -22,5 +22,5 @@ WORKDIR $WORKDIR
 
 RUN go get github.com/onsi/ginkgo/ginkgo
 
-COPY . .
-CMD ["make", "test"]
\ No newline at end of file
+#COPY . .
+#CMD ["make", "test"]
```
Comment (ArbelNathan) [inline build/ci/Dockerfile.unittest:26]: is it final? keeping it as a comment?
also missing line at end of file..

File: .gitignore [hunk 3]
```diff
@@ -80,3 +80,4 @@ tags
 .history
 # End of https://www.gitignore.io/api/go,vim,emacs,visualstudiocode
 /vendor/pkg/
+/bin/
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply .gitignore:83]: I don't think we should use any cache

File: build/ci/Dockerfile.unittest (hunk 2 above)
Comment (matancarmeli7) [reply build/ci/Dockerfile.unittest]: done

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (matancarmeli7) [reply Makefile]: > also @matancarmeli7 I think [CURDIR](https://github.com/IBM/ibm-block-csi-driver/blob/10ec5e1357c01d2a373cc31b66961f07136de1a6/Makefile#L52) is [preferred over PWD](https://stackoverflow.com/a/52447088) (are we missing a [convention alignment](https://github.com/IBM/ibm-block-csi-operator/blob/b9e04d5e50b9af905522354f52051f03ec417465/Makefile#L42-L46)?)
>
done

File: .gitignore (hunk 3 above)
[the remark this answers is not in the document]
Comment (matancarmeli7) [reply .gitignore]: done

File: hack/check-generated-manifests.sh [hunk 4]
```diff
@@ -20,7 +20,7 @@ project_dirname=ibm-block-csi-operator
 cd ..
 cp -r $project_dirname ./$project_dirname-copy
 cd $project_dirname-copy/
-hack/update-crds.sh
+hack/update-genrated-yamls.sh
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply hack/check-generated-manifests.sh:23]: then how about:
```suggestion
hack/update-yamls.sh
```

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply Makefile]: > the operator project is only on the go working dir..

@ArbelNathan is there any action item here for @matancarmeli7?

File: hack/check-generated-manifests.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply hack/check-generated-manifests.sh:23]: are the yamls updated be update-installer not generated?

[the remark this answers is not in the document]
Comment (oriyarde) [reply hack/check-generated-manifests.sh]: I'm not sure I follow which yamls exactly it updates

File: hack/update-installer.sh [hunk 5]
```diff
@@ -19,84 +19,6 @@
...
-    sed -i '' -e '$d' $TARGET_FILE
-    sed -i '' -e '$d' $TARGET_FILE
-elif [[ $os == "Linux" ]]
-then
-    sed -i '$d' $TARGET_FILE
-    sed -i '$d' $TARGET_FILE
-fi
+yes | cp deploy/installer/generated/kustomization.yaml .
+kustomize build -o deploy/installer/generated/ibm-block-csi-operator.yaml
+rm -rf kustomization.yaml
```
Comment (oriyarde) [inline hack/update-installer.sh:24]: why `-r`?

File: deploy/installer/generated/ibm-block-csi-operator.yaml [hunk 6]
```diff
@@ -1,18 +1,16 @@
----
 apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
     controller-gen.kubebuilder.io/version: v0.4.1
-  creationTimestamp: null
```
Comment (oriyarde) [inline deploy/installer/generated/ibm-block-csi-operator.yaml:7]: so why did we have it if we don't need it?

File: hack/check-generated-manifests.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (matancarmeli7) [reply hack/check-generated-manifests.sh]: it updates all the yamls under `config` directory:
1. when you do `make generate` you edit the crd yamls mainly, it to add new fields to the card, you add them in the type.go files and it edits the neccery crds
2. when you do `make manifests`, it edits the roles and the card yamls

File: Makefile [hunk 7]
```diff
@@ -55,26 +55,40 @@ TMP_DIR=$$(mktemp -d) ;\
...
 .PHONY: olm-validation
 olm-validation:
 	build/ci/olm_validation.sh
 
+.PHONY: build-unit-test-image
+build-unit-test-image:
+	docker build -f build/ci/Dockerfile.unittest -t operator-unittests .
+
+.PHONY: run-unit-test
+run-unit-test:
```
[the remark this answers is not in the document]
Comment (matancarmeli7) [reply Makefile:75]: done

File: hack/update-installer.sh [hunk 8]
```diff
@@ -19,84 +19,6 @@
...
-if [[ $os == "Darwin" ]]
-then
-    sed -i '' -e '$d' $TARGET_FILE
-    sed -i '' -e '$d' $TARGET_FILE
-elif [[ $os == "Linux" ]]
-then
-    sed -i '$d' $TARGET_FILE
-    sed -i '$d' $TARGET_FILE
-fi
+yes | cp deploy/installer/generated/kustomization.yaml .
```
[the remark this answers is not in the document]
Comment (matancarmeli7) [reply hack/update-installer.sh:22]: to eliminate mistakes, if someone creates it manually or something like this

File: deploy/installer/generated/ibm-block-csi-operator.yaml (hunk 6 above)
Comment (matancarmeli7) [reply deploy/installer/generated/ibm-block-csi-operator.yaml:7]: I don't know

File: deploy/installer/generated/kustomization.yaml [hunk 9]
```diff
@@ -0,0 +1,9 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
```
[the remark this answers is not in the document]
Comment (matancarmeli7) [reply deploy/installer/generated/kustomization.yaml:1]: it doesn't generate, so you think it better to move it one to `deploy/installer/` directory?

File: deploy/installer/generated/kustomization.yaml [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+kind: Kustomization
+
+resources:
+- config/crd/bases/csi.ibm.com_ibmblockcsis.yaml
+- config/manager/manager.yaml
+- config/rbac/role.yaml
+- config/rbac/role_binding.yaml
+- config/rbac/service_account.yaml
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply deploy/installer/generated/kustomization.yaml:9]: perhaps I wasn't clear.
the order in the operator yaml in this PR (service account before deployment) LGTM.
the diff on which I commented here has the "manager" before the service account - which to me seems unaligned

File: hack/update-installer.sh (hunk 5 above)
Comment (oriyarde) [reply hack/update-installer.sh]: it would be much easier to review these PRs if we prioritize CSI-3219

File: build/ci/run_unittests.sh [hunk 11]
```diff
@@ -16,5 +16,5 @@
 # limitations under the License.
 #
 
-docker build -f build/ci/Dockerfile.unittest -t operator-unittests .
-docker run --rm -t operator-unittests
+make build-unit-tests-image
+make run-unit-tests
```
Comment (oriyarde) [inline build/ci/run_unittests.sh:20]: is there any reason we want to require `make` installation in our CI?

Comment (matancarmeli7) [reply build/ci/run_unittests.sh:20]: isn't it a basic command in Linux environments?

Comment (oriyarde) [reply build/ci/run_unittests.sh:20]: is there any reason we want to require a Linux environment in our CI?

Comment (matancarmeli7) [reply build/ci/run_unittests.sh:20]: we not using windows as our CI

Comment (matancarmeli7) [reply build/ci/run_unittests.sh:20]: IMO using make commands makes the run more readable

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply Makefile]: talked to @matancarmeli7, he will test it on another working dir, preferably windows .

**Pull request merged**